Feature/allen/extrude with multiple vars (#18914)
visit-dav/visit: src/doc/using_visit/Operators/OperatorTypes/ExtrudeStacked_operator.rst

.. _ExtrudeStacked operator:

ExtrudeStacked operator
~~~~~~~~~~~~~~~~~~~~~~~

> The ExtrudeStacked operator uses one or more scalar fields on a 2D

Reviewer: We have a policy of a single sentence per line. Can you update accordingly, please.

Author: I do not understand this requirement. Would you provide an example.

Reviewer: The following, which has two sentences spread over four lines:
```
The ExtrudeStacked operator can be used to create 3D height field
plots by applying the operator to a plot that accepts scalar
values. The ExtrudeStacked operator is quite flexible because it can
extrude by more than one variable and displayed different ways.
```

should be as below, where each sentence is on one line, regardless of its length:
```
The ExtrudeStacked operator can be used to create 3D height field plots by applying the operator to a plot that accepts scalar values. 
The ExtrudeStacked operator is quite flexible because it can extrude by more than one variable and displayed different ways.
```

This requirement is documented in our [Contributing To Documentation](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/ContributingToDocumentation/index.html) with an expanded explanation related to [line-length](https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/ContributingToDocumentation/index.html#contributing-line-length).

Sphinx will take care of the wrapping as needed.

Author: Thanks I understand now but in looking at some of the other operators and plots none of them have that formatting.  I also note that at least in emacs it does some auto formatting when the lines get too long. And being old school I display 80 characters so one sentence per line is hard to read because of the wrap around.

Reviewer: Yes, the policy was introduced after we had already created some of the content. It was decided that those files would be made to comply with the policy either when someone updated the file, or during a future code-sprint.
New files must comply.

Author: Note to self: Do not ever touch a rst file again :-)

mesh to extrude each node in the input mesh, resulting in a 3D height
field. The ExtrudeStacked operator can also extrude higher dimensional
meshes that have been sliced producing a 2D mesh. The resulting
extruded mesh can be displayed by the extrusion height or an index
(see: :numref:`Figure %s <extrude>`).

.. _extrude:

.. figure:: images/extrude.png

   ExtrudeStacked operator example: A simple structured grid extruded
   by a single scalar cell value.

Using the ExtrudeStacked operator
"""""""""""""""""""""""""""""""""

The ExtrudeStacked operator can be used to create 3D height field
plots by applying the operator to a plot that accepts scalar
values. The ExtrudeStacked operator is quite flexible because it can

> extrude by more than one variable and displayed different ways.

Reviewer: ```suggestion
extrude by more than one variable and display in different ways.
```

Author: Should have been: extrude by more than one variable and **be** displayed different ways.

The most useful feature of the ExtrudeStacked operator is its ability
to extrude a mesh using multiple arbitrary scalar variables. By
**default**, the ExtrudeStacked operator uses the plot's variable to
extrude the mesh. This only works when the plotted variable is a
scalar variable. When one applies the ExtrudeStacked operator to plots
that do not accept scalar variables, the ExtrudeStacked operator will

> fail unless a specific scalar variable is choosen int the **Add

Reviewer: ```suggestion
fail unless a specific scalar variable is chosen in the **Add
```

Author: Done

(see :numref:`Figure %s <extrudeWindowStacked>`).

.. _extrudeWindowStacked:

.. figure:: images/extrudeWindowStacked.png

   ExtrudeStacked operator window showing multiple variables.

As the ExtrudeStacked operator uses a scalar variable to extrude all
of the points in the mesh, one can set the minimum and maximum values
as well as scale for each scalar variable. This allows one to
eliminate data that might otherwise cause the extruded mesh to be
stretched undesirably. To set the minimum and maximum values and scale
for the ExtrudeStacked operator, one first selects the variable, then
sets the **Min** or **Max** or **Scale** in the **ExtrudeStacked
attributes window** (see :numref:`Figure %s
<extrudeWindowScaled>`). The values will automatically be updated in
the variable table and then applied to the mesh (see :numref:`Figure
%s <extrudeScaled>`).

(see :numref:`Figure %s <extrudeWindowScaled>`).

.. _extrudeWindowScaled:

.. figure:: images/extrudeWindowScaled.png

   ExtrudeStacked operator window with the values scaled.

.. _extrudeScaled:

.. figure:: images/extrudeScaled.png

   ExtrudeStacked operator used to create a scaled plot using one

>    variable. The left image used node based scalar values used to

Reviewer: ```suggestion
   variable. The left image used node based scalar values to
```

Author: Done

The ExtrudeStacked operator uses one or more scalar fields on a 2D mesh to extrude each node in the input mesh, resulting in a 3D height field.
The ExtrudeStacked operator can also extrude higher dimensional meshes that have been sliced producing a 2D mesh.
The resulting extruded mesh can be displayed by the extrusion height or an index (see: :numref:`Figure %s <extrude>`).

.. _extrude:

.. figure:: images/extrude.png

   ExtrudeStacked operator example: A simple structured grid extruded by a single scalar cell value.

Using the ExtrudeStacked operator
"""""""""""""""""""""""""""""""""

The ExtrudeStacked operator can be used to create 3D height field plots by applying the operator to a plot that accepts scalar values.
The ExtrudeStacked operator is quite flexible as it can extrude by more than one variable and be displayed different ways.

The most useful feature of the ExtrudeStacked operator is its ability to extrude a mesh using multiple arbitrary scalar variables along an arbitrary **Extrusion axis**.
By **default**, the ExtrudeStacked operator uses the plot's current variable to extrude the mesh.
This only works when the plotted variable is a scalar variable.

> When one applies the ExtrudeStacked operator to plots that do not accept scalar variables, the ExtrudeStacked operator will
> fail unless a specific scalar variable is chosen int the **Add Variable** variable menu in the **ExtrudeStacked attributes window** (see :numref:`Figure %s <extrudeWindow>`).

Reviewer: ```suggestion
When one applies the ExtrudeStacked operator to plots that do not accept scalar variables, the ExtrudeStacked operator will fail unless a specific scalar variable is chosen in the **Add Variable** variable menu in the **ExtrudeStacked attributes window** (see :numref:`Figure %s <extrudeWindow>`).
```

As the ExtrudeStacked operator uses a scalar variable to extrude all of the points in the mesh, one can set the minimum and maximum values as well as scale for each scalar variable.
This allows one to eliminate data that might otherwise cause the extruded mesh to be stretched undesirably.
To set the minimum and maximum values and scale for the ExtrudeStacked operator, one first selects the variable, then sets the **Min** or **Max** or **Scale** in the **ExtrudeStacked attributes window** (see :numref:`Figure %s <extrudeWindowScaled>`).
The values will automatically be updated in the variable table and then applied to the mesh (see :numref:`Figure %s <extrudeScaled>`).

(see :numref:`Figure %s <extrudeWindowScaled>`).

.. _extrudeWindowScaled:

.. figure:: images/extrudeWindowScaled.png

   ExtrudeStacked operator window with the values scaled.

.. _extrudeScaled:

.. figure:: images/extrudeScaled.png

   ExtrudeStacked operator used to create a scaled plot using one variable.
   The left image used node based scalar values to extrude the mesh, the right image scaled the scalar values by two.
   Both are displayed by the node based scalar values (see the next section).

> It is possible to **Display by** the resulting field via the **Index** of the varible in the stack or by the scalar values used to extrude the mesh.

Reviewer: ```suggestion
It is possible to **Display by** the resulting field via the **Index** of the variable in the stack or by the scalar values used to extrude the mesh.
```

.. _ExtrudeStacked operator:

ExtrudeStacked operator
~~~~~~~~~~~~~~~~~~~~~~~

The ExtrudeStacked operator uses a scalar field on a 2D mesh to

> extrude each node in the input mesh, resulting in a 3D mesh. The The

Reviewer: ```suggestion
extrude each node in the input mesh, resulting in a 3D mesh. The
```

Author: done

ExtrudeStacked operator can also extrude plots whose input data was
produced from higher dimensional data that has been
sliced. Furthermore, the ExtrudeStacked operator allows you to display
multiple scalar fields in a single plot such as when a Pseudocolor

> plot of scalar variable A is extrudes by scalar variable B (see:

Reviewer: ```suggestion
plot of scalar variable A is extruded by scalar variable B (see:
```

Author: Done

:numref:`Figure %s <extrude>`).

.. _extrude:

.. figure:: images/extrude.png

   ExtrudeStacked operator example: A simple structured grid extruded
   by the cell number.

Using the ExtrudeStacked operator
"""""""""""""""""""""""""""""""""

The ExtrudeStacked operator can be used to create 3D height field

> plots by applying operator to a plot that accepts scalar values. The

Reviewer: ```suggestion
plots by applying the operator to a plot that accepts scalar values. The
```

Author: Done

ExtrudeStacked operator is quite flexible because it can extrude by
more than one variable and coloring by the same variable.

Since the ExtrudeStacked operator uses a scalar variable to extrude
all of the points in the mesh, the ExtrudeStacked operator has a
number of controls related to scaling scalar data. For example, the
ExtrudeStacked operator allows you to artificially set minimum and
maximum values for the scalar variable so you can eliminate data that
might otherwise cause your extruded plot to be stretched undesirably

> in the Z direction. To set minimum and maximum values for the
> ExtrudeStacked operator, set the **Min** or **Max** check boxes in the
> **ExtrudeStacked attributes window** (see :numref:`Figure %s
> <extrudeWindow>`) and type new values into the adjacent text
> fields.

Reviewer: ```suggestion
in the Z direction. To set minimum and maximum values for a variable, make sure that variable is highlighted in the table, then enter the desired minimum or maximum in the **Min** or **Max** text edit boxes below the table. (see :numref:`Figure %s
<extrudeWindow>`)  The table entries will be updated to the values entered.
```

The Min/Max aren't check boxes.
Also, I think it would be better if the example figure had multiple variables listed in the table, and min or max changed for the hightlighted variable to show that entering the min/max updates what is shown in the table, too.

Author: As I had an example with a scaled mesh I used that as an example for window.

Variable** variable menu in the **ExtrudeStacked attributes window**.
(see: :numref:`Figure %s <extrudeStacked>`).

.. _extrudeStacked:

.. figure:: images/extrudeStacked.png

   ExtrudeStacked operator used to created a stacked plot using two
   variables. The left image is colored based on scalar values, the
   right image is color based on the variable index.

Changing extrusion scaling
""""""""""""""""""""""""""

The ExtrudeStacked operator uses a scalar variable's data values as
the Z component when converting a mesh's 2D coordinates into 3D
coordinates. When the scalar variable's data extents are small/large
relative to the mesh's X and Y extents then you often get what appears
to be a flat 2D version or line in 3D space. It is sometimes necessary
to scale the scalar variable's data extents relative to the spatial
extents in order to produce a visualization where the Z value differs
noticeably. If you want to exaggerate the Z values that the scalar
variable contributes to make differences more obvious, you can click
on the **Scale Factor** check box in the **ExtrudeStacked attributes
window**. (see: :numref:`Figure %s <extrudeScaled>`).

> 

Reviewer: Please update this when you make the change to have scale apply to individual variables. You will need a new image, too.

Author: Revised
